test/clientserver_test.go

```diff
@@ -1,50 +1,270 @@
+}
+
+func (s testStatistic) Variance() float64 {
+	mean := s.Mean()
+	return s.totalSquared/float64(s.count) - mean*mean
+}
+
+func (s testStatistic) Stddev() float64 {
+	return math.Sqrt(s.Variance())
+}
+
+func (s testStatistic) MaxDeviation() float64 {
+	mean := s.Mean()
+	return max64(math.Abs(s.max-mean), math.Abs(s.min-mean))
+}
+
+func (s testStatistic) DeviatesBy(numStddev float64) bool {
+	return s.MaxDeviation()/s.Stddev() > numStddev
+}
+
+func TestClientServer(t *testing.T) {
+
+	var numIter []int = []int{1e3, 1e4, 1e5}
+	avgLatency := make([]time.Duration, len(numIter))
+	deltaGo := make([]int, len(numIter))
+	deltaMem := make([]uint64, len(numIter))
+	for i, n := range numIter {
+		avgLatency[i], deltaGo[i], deltaMem[i] = testClientServer(t, n)
+	}
+
+	// compute means and standard deviations of all statistics
+	var avgLatencyStat, deltaGoStat, deltaMemStat testStatistic
+	for i := range numIter {
+		avgLatencyStat.Add(float64(avgLatency[i]))
+		deltaGoStat.Add(float64(deltaGo[i]))
+		deltaMemStat.Add(float64(deltaMem[i]))
+	}
+
+	// each test instance executes with a different number of iterations (1e3, 1e4, 1e5).
+	// for each iteration, we measure three statistics:
+	//	- latency of average iteration (i.e. an rpc network call)
+	//	- excess/remaining number of goroutines after the test instance runs
+	//	- excess/remaining allocated memory after the test instance runs
+	// we then verify that no statistic regresses beyond 2 standard deviations across the different test runs.
+	// this ensures that none of the statistics grow with the increase in test iterations.
+	// in turn, this implies that there are no leakages of memory or goroutines.
+	const deviationFactor = 2
+	if avgLatencyStat.DeviatesBy(deviationFactor) {
+		t.Errorf("average latency is not stable")
+	}
+	if deltaGoStat.DeviatesBy(deviationFactor) {
+		t.Errorf("allocated goroutines count is not stable")
+	}
+	if deltaMemStat.DeviatesBy(deviationFactor) {
+		t.Errorf("allocated memory is not stable")
 	}
-	fmt.Println(infos)
 }
 
-func testFindProvidersAsyncFunc(key cid.Cid, ch chan<- client.FindProvidersAsyncResult) error {
-	ma := multiaddr.StringCast("/ip4/7.7.7.7/tcp/4242/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N")
-	ai, err := peer.AddrInfoFromP2pAddr(ma)
+func snapUtilizations() (numGoroutines int, alloc uint64) {
+	runtime.GC()
+	time.Sleep(time.Second)
+	var ms runtime.MemStats
+	runtime.ReadMemStats(&ms)
+	return runtime.NumGoroutine(), ms.Alloc
+}
+
+const (
+	testPeerID   = "QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N"
+	testPeerAddr = "/ip4/7.7.7.7/tcp/4242"
+)
+
+var (
+	// provider record
+	testMultiaddr = multiaddr.StringCast(testPeerAddr)
+	testAddrInfo  = &peer.AddrInfo{
+		ID:    peer.ID(testPeerID),
+		Addrs: []multiaddr.Multiaddr{testMultiaddr},
+	}
+	// IPNS
+	testIPNSID     []byte
+	testIPNSRecord []byte
+)
+
+// TestMain generates a valid IPNS key and record for testing purposes.
+func TestMain(m *testing.M) {
```

Comment:
@aschmahmann This code here doesn't seem to generate a correct pair of IPNS key and record. Can you advise what is wrong?

Reply:
The record is invalid because it's expired (you passed an EOL of `time.Now()`, note that the 1hr you passed was the TTL/how long the author thinks it's safe to cache records for as with DNS).

However, some issues in the code in the implementation here made surfacing this error to you more difficult (i.e. the only way you would've noticed the error was by reading an Info log). After these are fixed we should probably add a test for the expired/invalid records just to make sure we've bubbled the errors up correctly (i.e. to cover any changes we're making)